### ISSUE TITLE
[UX] Core's Bulk Operations: Select all items on all pages

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3798]
This is part of #3797

On a new site, I was plagued with 100's of new dodgy user accounts (because of #574). Because our core's Bulk Operations does not allow me to select all users from the many pages, my options were to either do a direct db query, use phpmyadmin, or manually repeat the process of selecting/deleting/confirming for each of the 12's of pages. This is not good UX, and the 7.x version of VBO included an option to select all items across all pages:

![Screen Shot 2019-05-20 at 9 12 38 am](https://user-images.githubusercontent.com/2423362/57989557-8d64cd00-7adf-11e9-89ba-ac25b8850e81.png)

Here's what that option allows:

![Screen Shot 2019-05-20 at 9 49 46 am](https://user-images.githubusercontent.com/2423362/57989909-b63b9100-7ae4-11e9-872b-65b0ac5a2a77.png)

---

Advocate: @cellear

**Comment by BWPanda** [comment]
Yes! I've missed having this in BD too.

**Comment by klonos** [comment]
...at least we have admin_views in core. I was able to edit the view, change the pager settings to show all items, then was able to select and delete all, then revert the pager settings back to default. Not a viable solution, and can be a performance issue on larger sites; just wanted to put the workaround here for future reference 😉

**Comment by herbdool** [comment]
We'll have to add functionality to https://github.com/backdrop/backdrop/blob/4426b2be634baa40dab541b2c6a248fe6b1ad6c6/core/modules/views/handlers/views_handler_field_bulk_form.inc (so I don't forget).

**Comment by stpaultim** [comment]
Needs "Feature Request" tag to show up on survey.

**Comment by argiepiano** [comment]
I'm going to change the title of this issue, as this refers to core's Bulk Operations, not to be confused with VBO (Views Bulk Operations) which has already been ported to BAckdrop and is available as a contrib module.

This issue/request specifically applies to core's Bulk Operation, not the Backdrop contrib Views Bulk Operations. The contrib VBO already has this ability, as @klonos pointed out in the OP.

**Comment by quicksketch** [comment]
This would make a great feature for 1.33.0. Although there is no PR or an advocate, this ability has a lot of prior implementation in contrib and in Drupal that we can pull from. Marking this as a milestone candidate for better visibility.

**Comment by quicksketch** [comment]
@cellear Volunteered during the dev meeting today to be the advocate for this issue. Thank you Luke!

**Comment by argiepiano** [comment]
@cellear, @quicksketch, here's a PR that adds the "Select all rows in view" functionality to views' core Bulk Form.

- Most of the js is borrowed from contrib's VBO, but modified, since core's BF uses tableselect functionality
- This new functionality only works for tables. This limitation is there because of the way BF was designed: since it uses tableselect to add the checkbox on the table header, there is no way to select all rows in the shown page when the view uses other display types. This _could_ be fixed, but I feel this is beyond this issue - in other words, BF would need a makeover pre-PR to include a checkbox to select all unordered-listed rows.
- Similar to contrib's VBO, the mechanism to select all rows in all pages is: creating a new view object to retrieve all records (without a pager), and using those entity IDs to send to the action. There is no easy alternative here, since we want to use exactly the same arguments and filters as the current view.

Please test and let me know. Happy Holidays!

Edit: unrelated test failure - the usual one but in php 8.1 this time.

**Comment by robertgarrigos** [comment]
Just tested. It works as expected. This is a great Xmas gift!

**Comment by argiepiano** [comment]
Just pushed a few more changes:
- I added a theming function for the markup for "select all". This mirrors what's in VBO, and will make it easier for site builders to override that theming function to modify the markup
- More importantly: I added the functionality for "Select all items on this page" and "Select all items in all pages" for **non-table** views. This functionality (specifically "Select all items on this page") was missing in core's Bulk Operations

A question: the new functionality for both tables and non-tables is enabled by default, meaning that existing views with core Bulk Operations will have that enabled. Is this OK? I could disable it in the handler definitions, but this would mean that, for new views that use that field, this functionality will be disabled. A third option might be to make it enabled by default, but then create an update hook that looks through ALL views in a site, and sets this functionality as disabled if there is a Bulk Operations field. I'm not sure this is worth it - would there be any reason that this functionality might be disabled?

**Comment by argiepiano** [comment]
I've added an automated test:
1. Creates a total of 20 nodes
2. Creates a view with a bulk_form field to unpublish nodes
3. Adds a pager for 10 nodes per page
4. Selects all items in all pages (new functionality)
5. Unpublishes those nodes
6. Checks that there are no published nodes left in the site

Disclaimer: this test (but not the functionality) was written with Copilot's help (it required several prompts and some manual modifications).

**Comment by yorkshire-pudding** [comment]
I plan to test this and look at the code later.  This would be a very welcome addition to core, and I use it regularly within contrib VBO

> A question: the new functionality for both tables and non-tables is enabled by default, meaning that existing views with core Bulk Operations will have that enabled. Is this OK? I could disable it in the handler definitions, but this would mean that, for new views that use that field, this functionality will be disabled. A third option might be to make it enabled by default, but then create an update hook that looks through ALL views in a site, and sets this functionality as disabled if there is a Bulk Operations field. I'm not sure this is worth it - would there be any reason that this functionality might be disabled?

I think this should be ok to be enabled by default. If this mirrors contrib VBO the select all is not forced, so there should be no unforeseen circumstances.

**Comment by yorkshire-pudding** [comment]
Thanks @argiepiano - I have tested this and it works well.
One question - is it feasible to display the count of the records selected for the non-table views? Not a deal-breaker if not.

I've reviewed the code and the functionality all makes sense to me.  There are a few minor actions to address.

**Comment by argiepiano** [comment]
Thanks, @yorkshire-pudding! I've made those changes:
- Namespacing and css classes now include `bulk_form` instead of `bf`
- Added the number of items per page and in view for non-tables

Additionally I made the label for non-table checkboxes clickable (by adding an `#id` to them, since that wasn't added as that particular build array never went through a form process callback). Ready for more testing.

**Comment by argiepiano** [comment]
Checking on those test failures.

**Comment by argiepiano** [comment]
OK, all tests passing, ready for testing and code review.

**Comment by yorkshire-pudding** [comment]
Thanks @argiepiano . The addition of the number of items for non-tables is good and works well when filters are adjusted.
All still works well.

RTBC

**Comment by quicksketch** [comment]
Woo! Thank you @argiepiano! A great holiday present indeed! I'm on vacation at the moment but I'll have time early next week to fully review. Initial impression is good and I'm glad to see automated test coverage is already included.

**Comment by olafgrabienski** [comment]
> A question: the new functionality for both tables and non-tables is enabled by default, meaning that existing views with core Bulk Operations will have that enabled. Is this OK?

Sorry for the late reply. I think the new option is great, but I would prefer it not to be automatically enabled, as it's very far-reaching. For example, it allows editors to delete all posts on a website at once. In my opinion, such an option should only be available based on a conscious decision by an admin and/or the site owner.

**Comment by herbdool** [comment]
@olafgrabienski before this PR it would allow an editor to select all posts on a page and delete them in batches. To me that already would allow them to delete all posts on the site. It would just require them to page through.

**Comment by olafgrabienski** [comment]
@herbdool You're right, and I find the current behavior also a bit risky. That's why I tend to disable the option to bulk delete content for editors at all. (Also because Backdrop doesn't have a trash bin.) However, if an editor uses it by accident, the option to delete _all content_ at once makes a big difference to the current behavior.

**Comment by argiepiano** [comment]
It would be easy to make the "select all rows in view" disabled by default, if that's the desired behavior. That would allow admins to manually enable it for new views, or to edit old views to enable. Honestly I find the lack of "select all in view" weird - if the admin adds a bulk operation to a view, that's what we always expected - to be able to do things in bulk - the fact that the view has a pager is just an accident that currently prevents the full power of bulk operations. That ("select all in view" enabled) was the default in D7.

**Comment by yorkshire-pudding** [comment]
> [@herbdool](https://github.com/herbdool) You're right, and I find the current behavior also a bit risky. That's why I tend to disable the option to bulk delete content for editors at all. (Also because Backdrop doesn't have a trash bin.) However, if an editor uses it by accident, the option to delete _all content_ at once makes a big difference to the current behavior.

@olafgrabienski - there are probably some soft delete / trash bin /recycle bin modules for D7 that could be ported to fill that gap.

I normally restrict permissions for views with bulk operations and there is a confirm step so I've never accidentally deleted more than I planned.  The key is probably having the right other filters to target the selection appropriately.

**Comment by herbdool** [comment]
I still think it should be enabled by default because I consider it a usability improvement for those who already have permission to bulk delete/update. If not then we've got a discoverability issue where most site admins just won't know it's an option.

**Comment by olafgrabienski** [comment]
@argiepiano @yorkshire-pudding @herbdool – Thanks for the feedback! I see all your points and remain a bit skeptical at the same time. I'll reconsider the topic and will try to come up with a suggestion which respects different perspectives.

**Comment by quicksketch** [comment]
> However, if an editor uses it by accident, the option to delete all content at once makes a big difference to the current behavior.

I think because we have a warning and confirmation page for Delete already, this should be pretty obvious the amount of deletion that is going to occur. In the Tugboat sandbox for example:

<img width="741" height="394" alt="Image" src="https://github.com/user-attachments/assets/86a0bba6-de0f-4076-8005-3fbc30d7f974" />

I don't think we have any special logic in there for a large number of items. So if you tried to delete 10,000 nodes or something, the page would literally list every single node about to be deleted. I think that's clear enough, though it will probably cause other issues like PHP timeouts for that many deletions.

**Comment by argiepiano** [comment]
> I don't think we have any special logic in there for a large number of items. So if you tried to delete 10,000 nodes or something, the page would literally list every single node about to be deleted. 

True! This is a problem with `node_multiple_delete_confirm()`, whose use is not limited to the Bulk Form (it has it's own menu path). I could modify this as part of this PR or open a new issue. Any thoughts about this, @quicksketch?

**Comment by quicksketch** [comment]
> I could modify this as part of this PR or open a new issue. Any thoughts about this, @quicksketch?

I'm keen on getting this in, so I'd like to separate out enhancements to a follow-up.

**Comment by argiepiano** [comment]
I'll open a new issue.  EDIT: https://github.com/backdrop/backdrop-issues/issues/7034

I'm thinking that the best way to handle batch processing is to mirror what VBO does. `action_execute()` (or `action_do()` in D7) do not inherently do batch process of the actions. Instead this is handled by VBO - so it should be handled by the Bulk Form views handler in Backdrop's case. I agree this can be tackled later. I can tackle this but it will take me a few days. In the meantime, if this PR looks OK, it could be merged so that it makes it into the next release.

**Comment by quicksketch** [comment]
I filed another follow-up to improve the namespacing and variables a bit. I didn't want to hold up a working RTBC issue but I think we could refresh the naming to align better with core Views module: https://github.com/backdrop/backdrop-issues/issues/7038

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/5178 into 1.x for 1.33.0. Thank you @argiepiano for coming in to make this happen! Thank you @herbdool, @olafgrabienski, @yorkshire-pudding, and @robertgarrigos for testing and review!
